File: cpp/leetcode/longest_substring_without_repeating_characters.cc
#define _USE_MATH_DEFINES

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
   public:
    int lengthOfLongestSubstring(string s) {
        map<char, int> visited;
        int ans = 0, start = 0;
        for (int i = 0; i < s.size(); i++) {
            if (visited.find(s[i]) != visited.end() && visited[s[i]] >= start) {
                ans = max(ans, i - start);
                start = visited[s[i]] + 1;
            }
            visited[s[i]] = i;
        }
        ans = max(ans, (int)s.size() - start);
        return ans;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cout << (new Solution())->lengthOfLongestSubstring("abcabcbb") << "\n";
    cout << (new Solution())->lengthOfLongestSubstring("abba") << "\n";

    return 0;
}

File: rust/src/leetcode/longest_substring_without_repeating_characters.rs
use std::cmp::max;
use std::collections::HashMap;

struct Solution;
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut visited = HashMap::new();
        let mut ans = 0;
        let mut start = 0;
        for (i, char) in s.chars().enumerate() {
            if let Some(&index) = visited.get(&char) {
                if index >= start {
                    ans = ans.max(i as i32 - start);
                    start = index + 1;
                }
            }
            visited.insert(char, i as i32);
        }
        max(ans, s.len() as i32 - start)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_length_of_longest_substring() {
        println!("{}", Solution::length_of_longest_substring(String::from("abcabcbb")));
        println!("{}", Solution::length_of_longest_substring(String::from("abba")));
        println!("{}", Solution::length_of_longest_substring(String::from(" ")));
    }
}
